### DISCUSSION TITLE
Error when compiling with njs

### DISCUSSION THREAD

**Discussion by drmad** [nginx/unit#1026] [Support] [answered]
Hi

Compiling unit 1.31.1 with njs 0.8.2 fails with:

```
src/nxt_js.c:75:5: error: excess elements in struct initializer [-Werror]
   75 |     nxt_js_module_loader,
      |     ^~~~~~~~~~~~~~~~~~~~
src/nxt_js.c:75:5: note: (near initialization for ‘nxt_js_ops’)
src/nxt_js.c:76:5: error: excess elements in struct initializer [-Werror]
   76 |     NULL,
      |     ^~~~
src/nxt_js.c:76:5: note: (near initialization for ‘nxt_js_ops’)
cc1: all warnings being treated as errors
make: *** [build/Makefile:1419: build/src/nxt_js.o] Error 1
```

Any advice? Thanks!

**Comment by drmad** [comment]
unit `njs_js_ops` is defined here:

https://github.com/nginx/unit/blob/d9f5f1fb741109cc232cedd3574aa587626789c1/src/nxt_js.c#L72-L77

has 4 elements, but in `njs.h`, typedef `njs_vm_ops_t` only has 2:

https://github.com/nginx/njs/blob/2c937050a4589bbc196db334fef22e6de772dd49/src/njs.h#L216-L219

Should I create an issue about this?

**Comment by hongzhidao** [comment, marked as answer]
Hi @drmad,
Good catch, thanks. We've noticed it and now it's better to use njs 0.8.1 for Unit 1.31.
We'll release 1.32, maybe next month, then we'll support njs 0.8.2.

Btw, I created a related issue.
https://github.com/nginx/unit-docs/issues/72

**Reply by drmad** [reply]
Nice, thanks.

Perhaps you shoud add the warning in https://unit.nginx.org/howto/source/#source-njs also

**Comment by drmad** [comment]
Meanwhile, I removed lined 73 and 74 from `unit/src/nxt_js.c` and 41 and 42 from `unit/src/nxt_script.c`. Compilation passed, and it's working apparently well now.

**Reply by hongzhidao** [reply]
Great job.
Would you mind creating a PR? It belongs to the 1.32 roadmap, but we haven't started it.

**Comment by ac000** [comment]
If that's the only differences that could be handled with some `#ifdef`'s
